_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -1,131 +1,131 @@
+  as well as the predictions, order quantities, and prices themselves.
+
+  ### Downloading data from a database to a dataset
+
+  One typical use case that we have is downloading large amounts of data from a database into a
+  dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is
+  an extremely simple: a lot of parallel nodes downloading pieces of data into a blob and one
+  reduction node writing the dataset metadata (which is basically a dictionary what data can be found in which blob).
+
+  ### Doing computations on a dataset
+
+  A lot of our algorithms for machine learning, forecasting, and optimization work on datasets.
+  A source dataset is read, computations are performed on the data, and the result is written to
+  a target dataset.
+  In many cases, the layout of the source dataset (the partitioning, i.e., what data resides in which blob)
+  is used for parallelization. This means the algorithms work independently on the individual
+  blobs of the source dataset. Therefore our respective Dask graphs again look very simple, with parallel nodes each performing
+  the sequential operations of reading in the data from a source dataset blob, doing some computation on it,
+  and writing it out to a target dataset blob. Again, there is a final reduction node writing the target
+  dataset metadata.
+
+  ### Dataset maintenance
+
+  We also use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
+  Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
+  references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
+  In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
+  In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.
+
+  ### Repartitioning a dataset
+
+  Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+  the case that we have the data available in a suitable partitioning, for instance, if the data results
+  from a computation that used a different partitioning. In this case, we have to repartition the data.
+  This works by reading the dataset as a Dask distributed dataframe (ddf), repartitioning the ddf using
+  network shuffle, and writing it out again to a dataset.
+
+  ## Dask cluster setup at Blue Yonder
+
+  We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+  of our domain and partly historically grown. We are running a large number of individual clusters
+  that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+  setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+  of upstream pull requests.
```

Comment:
I think mentioning stuff like Semaphore implemented by us on distributed would be very beneficial. Would be helpful to make it analogous to a similar feature in PySpark (if it has)

Reply:
I agree that it's a good idea to include references to PRs we've done, but I think the semaphore is not an example of an edge case we have run into because of our cluster setup, but rather a missing feature we needed. But maybe we could add a separate paragraph on that?